subprojects/packagefiles/kafel/meson.build

```
@@ -0,0 +1,72 @@
+project('kafel', 'c',
+  version: '2023.10.10',
+  default_options: [
+    'c_std=gnu11',
+  ])
+
+flex = find_program('flex')
+bison = find_program('bison')
+pkgconf = import('pkgconfig')
+
+run_command(flex, '--outfile=lexer.c', 'src/lexer.l', check: true)
+run_command(bison, '-o=parser.c', 'src/parser.y', check: true)
```

You don't want to use `run_command` for this, this is what `custom_target` is for:
```meson
lexer = custom_target(
   command : [flex, '--outfile', '@OUTPUT0@', '--header-file='@OUTPUT1@', '@INPUT@'],
   input : 'src/lexer.l',
   output : ['@BASENAME@.c', '@BASENAME@.h'],
)
parser = custom_target(
  command : [bison, '-o', '@OUTPUTo@', '--defines=@OUTPUT1@', '@INPUT@'],
  input : 'src/parser.y',
  output : ['@BASENAME@.c', '@BASENAME@.h'],
)
```

---

So, I had already tried it that way, but it didn't work, so I resorted to using run_command. Many people who use flex and bison seem to face the same issue and end up resorting to this method.

---

With this method it simply does not generate the .c files.

---

It won’t generate the c file if you don’t add the parser variable to the target, because custom targets are not built by default.

---

I am adding it, and it still doesn't work.

---

Here's an example of how to do it in real working code: https://github.com/dcbaker/meson-plus-plus/blob/main/src/frontend/meson.build

---

So, I had already tried using `custom_target` before, but without success.
It fails to generate the files at the exact moment they are needed, resulting in file not found errors. I attempted to fix this but was unable to resolve the issue.
However, after researching, I found an approach using `run_command`, which is used in similar contexts to mine.
Some examples:

- https://github.com/postgrespro/jsquery/blob/3073aebabdac58b082e95a63299216ca363feec6/meson.build#L22

- https://github.com/srbarton43/programmers_calculator/blob/bbed747521328c6b31168118c178d54fb19b3b9b/src/meson.build#L12

---

Neither of those projects are doing the right thing. There are two problems with the run_target approach you're taking,
1) it writes files into the source directory. Meson generally promises not to do this and it leads to issues, many cmake projects specifically go out of their way to turn off in-tree builds
2) the flex and bison files get written at configure time instead of build time, which means that their inputs are not properly tracked (so among other things ninja clean cant erase them), and to regenerate them a user needs to *manually* re-run meson, since Meson doesn't know those command write files

The critical thing here is that the output of the assigned variable from the custom_target needs to be in the sources list of the target that requires it:
```meson
x = custom_target(...)
library(
  'foo',
  'foo.c', 'lib..c', x,
)
```
That is how Meson knows that `libfoo` depends on the output of `x`, and properly orders it so that `x` is generated before any files in `libfoo` start compiling. the Meson++ example I pointed you to demonstrates how to write this as a `custom_target`, which is the right way to do this.

`run_target` isn't for codegen, it's for things like running linters or code formatters.

---

Yes, I tried this method as well, but the problem persists. The files are still being generated in another folder, causing the error.
```
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -C /home/walt/Desktop/Trabalho/kafel/emilua/build
ninja: Entering directory `/home/walt/Desktop/Trabalho/kafel/emilua/build'
[31/365] Compiling C object subprojects/kafel-20231004/libkafel.so.p/meson-generated_.._lexer.c.o
FAILED: subprojects/kafel-20231004/libkafel.so.p/meson-generated_.._lexer.c.o 
cc -Isubprojects/kafel-20231004/libkafel.so.p -Isubprojects/kafel-20231004 -I../subprojects/kafel-20231004 -I../subprojects/kafel-20231004/include -I../subprojects/kafel-20231004/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu11 -O0 -g -fPIC -MD -MQ subprojects/kafel-20231004/libkafel.so.p/meson-generated_.._lexer.c.o -MF subprojects/kafel-20231004/libkafel.so.p/meson-generated_.._lexer.c.o.d -o subprojects/kafel-20231004/libkafel.so.p/meson-generated_.._lexer.c.o -c subprojects/kafel-20231004/lexer.c
cc1: fatal error: subprojects/kafel-20231004/lexer.c: No such file or directory
compilation terminated.
[34/365] Compiling C object subprojects/kafel-20231004/libkafel.so.p/src_kafel.c.o
FAILED: subprojects/kafel-20231004/libkafel.so.p/src_kafel.c.o 
cc -Isubprojects/kafel-20231004/libkafel.so.p -Isubprojects/kafel-20231004 -I../subprojects/kafel-20231004 -I../subprojects/kafel-20231004/include -I../subprojects/kafel-20231004/src -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu11 -O0 -g -fPIC -MD -MQ subprojects/kafel-20231004/libkafel.so.p/src_kafel.c.o -MF subprojects/kafel-20231004/libkafel.so.p/src_kafel.c.o.d -o subprojects/kafel-20231004/libkafel.so.p/src_kafel.c.o -c ../subprojects/kafel-20231004/src/kafel.c
../subprojects/kafel-20231004/src/kafel.c:28:10: fatal error: lexer.h: No such file or directory
   28 | #include "lexer.h"
      |          ^~~~~~~~~
compilation terminated.
```

---

Any tips?

---

Can you show me what your updated meson.build looks like? even just throwing it on pastebin or a github gist would be fine.

---

https://pastebin.com/SRhvc3nw

---

The meson.build looks correct, and builds fine for me with the kafel git master when run as a root project:
```
git clone --depth 1 https://github.com/google/kafel.git 
<download pastebin to meson.build>
meson setup builddir
ninja -C builddir
```

And manually inspecting the build.ninja shows exactly what I'd expect, every .c -> .o step in libkafel depends on lexer.h and parser.h.

But it fails as a subproject...

And I'm totally stumped *why*. Everything says that it's correct, but then the actual files are dumped into the root of the build directory, not into the project directory.

I'm running out of time for today. But I'll look at this more later (might be tomorrow my time)

---

parser.y:

```
%output "parser.c"
%defines "parser.h"
```

This overrides all command line arguments passed to the `bison` command. They will be silently ignored, no diagnostics shall be offered.

---

From the upstream Makefile:

```make
lexer.h lexer.c: lexer.l
	flex $<

parser.h parser.c: parser.y
	bison $<
```

So this is the classic badly written homebrew Makefile nonsense. They'd rather encode arguments into the source file and just... not do build commands in their Makefile.

---

This could be handled currently, by writing a script that is used to run bison/flex after `cd`ing to the output directory. Or by patching the upstream sources. There is no `workdir` argument for custom_target, though.

---

> The meson.build looks correct, and builds fine for me with the kafel git master when run as a root project:
> 
> ```
> git clone --depth 1 https://github.com/google/kafel.git 
> <download pastebin to meson.build>
> meson setup builddir
> ninja -C builddir
> ```
> 
> And manually inspecting the build.ninja shows exactly what I'd expect, every .c -> .o step in libkafel depends on lexer.h and parser.h.
> 
> But it fails as a subproject...
> 
> And I'm totally stumped _why_. Everything says that it's correct, but then the actual files are dumped into the root of the build directory, not into the project directory.
> 
> I'm running out of time for today. But I'll look at this more later (might be tomorrow my time)

Thanks, exactly that.

---

> parser.y:
> 
> ```
> %output "parser.c"
> %defines "parser.h"
> ```
> 
> This overrides all command line arguments passed to the `bison` command. They will be silently ignored, no diagnostics shall be offered.
> 
> Same for lexer.l:
> 
> ```
> %option outfile="lexer.c" header-file="lexer.h"
> ```

Damn, what now?
Do I have to edit them manually then? Is this allowed on Meson?

---

> This could be handled currently, by writing a script that is used to run bison/flex after `cd`ing to the output directory. Or by patching the upstream sources. There is no `workdir` argument for custom_target, though.

worked with:

``

---

Using `&&` in a custom_target command is officially a bug, it is not supposed to work but a part of the backend code that handles static libraries accidentally overrides it to make it work.

On the other hand:
> Damn, what now?
> Do I have to edit them manually then? Is this allowed on Meson?

subproject wraps support:

- diff_files: https://mesonbuild.com/Wrap-dependency-system-manual.html#diff-files
- copying the file over with a custom_target in the form of e.g. a sed or python script that deletes content dynamically
- submit a PR to upstream that fixes this in the long term

---

I used sed and fixed the issues, thanks!